Add pagination info in the query parameter

sadiqkhoja opened the pull request:
Closes getodk/central#1348

#### What has been done to verify that this works as intended?

Manual verification + added few tests.

#### Why is this the best possible solution? Were any other approaches considered?

Please see inline comments.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?

No

#### Does this change require updates to user documentation? If so, please file an issue [here](https://github.com/getodk/docs/issues/new) and include the link below.

No

#### Before submitting this PR, please make sure you have:

- [x] run `npm run test` and `npm run lint` and confirmed all checks still pass OR confirm CircleCI build passes
- [x] verified that any code or assets from external sources are properly credited in comments or that everything is internally sourced

sadiqkhoja committed:
Feature getodk/central#1348: Add pagination info in the query parameter

PR Feedback
+ Created paginationQueryRef composable
+ Removed pagination query parameter when navigating to the map view
+ Included skipSettingSnapshotFilter to the fetchChunk function

matthew-white commented:
A few quick thoughts from me!

1. Order of PRs

I think this PR will conflict with getodk/central#1434, just to the extent that I'll be adding an `EntityTableView` component for that. Should I go ahead and push a PR for that now? Or would it be easier to get this one merged first? I've been working on other aspects of entity mapping, so I'd be happy for this one to go first if it could get merged quickly.

2. Toggling to/from map view

If there are pagination-related query parameters, do those currently persist when toggling to map view? For example:

- Navigate to the table.
- Move to the second page of the table (`?page-number=2`).
- Move to map view.
    - I would expect the `page-number` query parameter to be removed at this point.
- Move back to table view.
    - I would expect to see the first page of the table.

3. Composable

How about we add a composable for the pagination-related query parameters? I'm thinking of something similar to `useDateRangeQueryRef()`. I could see such a composable being useful in the future beyond submissions and entities.

matthew-white commented on src/components/entity/list.vue:
```diff
@@ -187,10 +187,32 @@ export default {
+        if (!num || num < 0) return 0;
+        return num - 1;
+      },
+      toQuery: (value) => ({ 'page-number': value + 1 })
```
What do you think about removing the query parameter for the first page? Since the first page is the default — one less query parameter in the URL.

```suggestion
      toQuery: (value) => ({ 'page-number': value === 0 ? null : value + 1 })
```

sadiqkhoja commented on src/components/entity/list.vue:
```diff
@@ -43,7 +43,7 @@ except according to the terms contained in the LICENSE file.
         :top="pagination.size"
         :filter="odataFilter != null || !!searchTerm"
-        :total-count="dataset.dataExists ? dataset.entities : 0"/>
+        :total-count="dataset.dataExists && !pagination.page ? dataset.entities : 0"/>
```
I don't want to show "Loading first N entities...", with this change it will always display "Loading entiities..." when there is a pagination information in the query parameter on the first loading. Note that on change pages using pagination control doesn't show any loading message.

matthew-white replied:
It'd be great if we could resolve this comment before merging. If this PR will make several i18n messages irrelevant, I think it'd be best to remove them. That could be done in a follow-up PR, but it'd be nice to track the need to do so in a new issue.

sadiqkhoja commented on src/components/submission/table-view.vue:
```diff
@@ -137,6 +161,11 @@ const fetchChunk = (clear, refresh = false) => {
+      const lastPage = Math.max(0, Math.ceil(pagination.count / pagination.size) - 1);
+      if (pagination.page > lastPage) {
+        pagination.page = lastPage;
+        fetchChunk(false);
```
introduced `skipSettingSnapshotFilter` in fetchChunk

sadiqkhoja commented on src/composables/pagination-query-ref.js:
```diff
@@ -0,0 +1,41 @@
+export default (pageSizeOptions = [250, 500, 1000]) => {
+  const pageNumber = useQueryRef({
+    fromQuery: (query) => {
+      const num = Number.parseInt(query['page-number'], 10);
```
I using `page-number` for the readability sake, we can also go with just `page`.

matthew-white replied:
I definitely prefer `page` over `page-number`. It's much shorter, and I see it on many sites. `page` + `page-size` sounds like a nice combo to me. I don't have the strongest feelings about it though, we could run with `page-number` for now and see if anyone else on the team voices an opinion.

sadiqkhoja commented on src/composables/pagination-query-ref.js:
```diff
@@ -0,0 +1,41 @@
+    fromQuery: (query) => {
+      const num = Number.parseInt(query['page-number'], 10);
+      if (!num || num < 0) return 0;
+      return num - 1;
```
I guess page number starting from 1 is more acceptable than 0.

matthew-white commented on src/components/submission/table-view.vue:
```diff
@@ -88,12 +90,10 @@ const odataSelect = computed(() => {
-  const first = clear || refresh;
-  if (first) {
+  if (refresh) {
     pagination.page = 0;
```
I'm trying to think through the ramifications of `pagination.page = 0;` for a background refresh. Wouldn't that cause the pagination control to immediately change to the first page? But the actual table wouldn't change until the request completes. If that takes a while, I think it'd be confusing for the pagination control to say that it's on the first page. Or even worse, if the request results in an error, the pagination control would incorrectly think that it's on the first page.

If it's a background refresh, I feel like `pagination.page` should be set to `0` only once the response is received.

It kind of feels like that would be changing the current behavior of background refresh though. Right now, does it refresh the current page rather than returning to the first page? If so, I feel like changing that behavior should maybe go in its own PR. Then this PR could just be focused on query parameters.

sadiqkhoja replied:
I have tested on test.getodk.cloud and on pressing of refresh button it goes back to the first page.

What I have done now is that `$skip` would send 0 if `refresh` or `clear` is true otherwise use pagination options. And `pagination.page=0` happens only in the success response.

matthew-white commented on src/composables/pagination-query-ref.js:
```diff
@@ -0,0 +1,41 @@
+https://www.apache.org/licenses/LICENSE-2.0. No part of ODK Central,
+including this file, may be copied, modified, propagated, or distributed
+except according to the terms contained in the LICENSE file.
+*/
```
We don't need these file headers for new files anymore. 🥳 So this can be removed.

matthew-white reviewed (commented):
It's looking good!

Probably the main question I still have is where/when `pagination.page` is set to 0.

In some cases, I made non-essential suggestions. I know we want to get this PR merged soon, so I'm happy with us skipping some of those suggestions or filing follow-up issues about them.

sadiqkhoja committed:
PR Feedback:

introduced `page` arg in fetchChunk()
set page=0 in success response of the request
removed copyright header
removed default pageSizeOptions from pagination-query-ref
various changes to the tests

sadiqkhoja commented on src/components/submission/table-view.vue:
when I change the page number in the URL, I see whole page gets reloaded both in chrome and firefox, which sets the correct pagination options before request is made hence page itself shows the correct data. I hope I understood your comment correctly.

matthew-white replied:
Ah right, I forgot that's how it works. This might have been a concern under the old hash-based routing, but not anymore.

sadiqkhoja commented on src/components/submission/table-view.vue:
```diff
@@ -126,7 +131,10 @@ const fetchChunk = (clear, refresh = false) => {
 fetchChunk(true);
-watch([() => props.filter, () => props.deleted], () => { fetchChunk(true); });
+watch([() => props.filter, () => props.deleted], () => {
+  pagination.page = 0;
```
Since I have move `pagination.page=0` in the `then` of the request, it will retain the existing behavior of changing columns.

matthew-white commented on test/util/router.js:
I think these comments are out-of-date now:

- The mock router can now navigate.
- It's no longer read-only.

matthew-white commented on apps/central/test/util/router.js:
```diff
@@ -75,7 +75,10 @@ export const mockRouter = (location = undefined) => (container) => {
+    push: sinon.fake(),
+    replace: (loc) => {
+      currentRoute.value = router.resolve(loc);
+    },
```
I don't feel like it totally works to make this change on its own. For example, `$route` is currently set to a constant at install time under the assumption that it will never change. But that assumption won't be true anymore.

I'm thinking we could create a `Proxy` object that delegates to `currentRoute.value`. Something like:

```js
const routeProxy = new Proxy(currentRoute, {
  get: (route, prop) => route.value[prop]
});

app.config.globalProperties.$route = routeProxy;

app.provide(routeLocationKey, routeProxy);
```

matthew-white commented on test/util/router.js:
```diff
@@ -75,7 +75,10 @@ export const mockRouter = (location = undefined) => (container) => {
       app.component('RouterView', RouterViewStub);
     },
-    push: sinon.fake()
+    push: sinon.fake(),
```
What do you think about making `push()` the same as `replace()`? It feels a little unexpected that `replace()` will change the route, while `push()` won't. `push()` could remain a Sinon fake. So something like:

```js
push: sinon.fake(loc => {
  currentRoute.value = router.resolve(loc);
}),
```

I'm not sure how tests currently use `push()`, so maybe this would cause problems, but hopefully that'd work.

matthew-white reviewed (commented):
I haven't taken a look at all the changes yet, but I thought I'd leave some initial comments about `mockRouter()` and `replace()`. Related comment: https://github.com/getodk/central-frontend/pull/1417#discussion_r2936246221

sadiqkhoja committed:
make currentRoute in mockRouter a proxy

matthew-white commented on apps/central/test/components/submission/list.spec.js:
```diff
@@ -129,6 +129,25 @@ describe('SubmissionList', () => {
         });
     });
+
+    it('makes the correct request after refresh button is pressed on second page', () => {
```
Could you add a similar test for `EntityList`? It looks to me like `SubmissionList` and `EntityList` work slightly differently in this area.

matthew-white commented on src/components/entity/table-view.vue:
```diff
@@ -107,19 +112,21 @@ const fetchChunk = (clear, refresh = false) => {
 // For defineExpose()
 const exposedFetch = {
-  fetchData: (clear = true) => fetchChunk(clear, !clear),
+  fetchData: (clear = true) => fetchChunk(clear, clear ? 0 : pagination.page),
```
```suggestion
  fetchData: (clear = true) => fetchChunk(clear),
```

I'm thinking that the second argument here should always be 0 (or should be omitted). At https://github.com/getodk/central-frontend/pull/1417#discussion_r3065291791, you wrote:

> I have tested on test.getodk.cloud and on pressing of refresh button it goes back to the first page.

But to me, that doesn't seem to be what this line is doing. If `clear` is `false` (if the refresh button has been pressed), then the user will remain on their current `pagination.page`, not go back to the first page.

matthew-white commented:
I think this comment thread still needs to be resolved: https://github.com/getodk/central-frontend/pull/1417#discussion_r2566357871.

I glanced over all the discussion above, and with the exception of that thread and my review today, I didn't see any other open threads.

matthew-white reviewed (approved):
Looks great! 💯 I know we've been discussing this one for a while, but I think this PR has ended up in a really nice place. I think it says something that it now only modifies 8 files. I really like the unit tests of the `usePaginationQueryRef()` composable — they feel like they're at the right level.

There are still a few comments left to resolve. Nothing big though, so I'll go ahead and approve. @sadiqkhoja, feel free to merge once you've resolved those comments!

sadiqkhoja committed:
fix fetchChunk arg and updated a test

sadiqkhoja committed:
remove irrelevant messages from loading message component
